Naive implementation of sub block.

ephiepark opened the pull request:
Naive implementation of sub block. This seems to pass fuzz test.

ephiepark committed:
Adding targetCBlockSize param

ephiepark committed:
Factor out the logic to build sequences

ephiepark committed:
Naive implementation

Cyan4973 commented on lib/compress/zstd_compress.c:
```diff
@@ -385,6 +385,11 @@ ZSTD_bounds ZSTD_cParam_getBounds(ZSTD_cParameter param)
         return bounds;
 
+    case ZSTD_c_targetCBlockSize:
+        bounds.lowerBound = 0; // TODO probably need lower bound
```
Good point !
Yes, very likely, and it will have to be documented.

Not obvious to decide one.
For a limit, we don't need to force a certain level of efficiency.
We merely need a limit which avoids triggering weird side effects in the implementation, and does not constrain the user too much. So it's likely going to be small.

Probably something smaller than 100 bytes, like 50, or 64 bytes.
I have no specific preference. I just seems the right ballpark.

Cyan4973 commented on lib/compress/zstd_compress.c:
```diff
@@ -2462,6 +2478,21 @@ static int ZSTD_disableLiteralsCompression(const ZSTD_CCtx_params* cctxParams)
+    switch (cctxParams->targetCBlockSize) {
+    case 0:
+        return 0;
+    default:
```
nit: `return (cctxParams->targetCBlockSize != 0)` feels more straightforward.

Cyan4973 commented on lib/compress/zstd_compress.c:
```diff
@@ -2626,6 +2657,166 @@ ZSTD_compressSequences_internal(seqStore_t* seqStorePtr,
+    /* Prepare nextEntropy assuming reusing the existing table */
+    memcpy(nextHuf, prevHuf, sizeof(*prevHuf));
+
+    /* small ? don't even attempt compression (speed opt) */
```
This limit might not be the most adapted one in this context.

The "small size" shortcut is a good idea when cumulating both the huffman header and huffman compressed symbols.

When the huffman header is essentially "free", because it's re-used from previous block (or from dictionary), the limit becomes much lower. In [this code](https://github.com/facebook/zstd/blob/dev/lib/compress/zstd_compress.c#L1957), the limit is down to `6` when huffman header is implicit.

ephiepark replied:
Currently, I am not considering set_repeat to simplify problem. I think I should add back the reduced limit check when I add the logic for set_repeat back in. What do you think?

Cyan4973 replied:
OK, so these are statistics for the superblock.
It makes sense, since there is no need (yet) to rebuild entropy tables between sub-blocks.
And, yes, in this case, your suggestion is completely reasonable.

Yet, since I did not understood it was statistics for the superblock,
it probably implies the naming can be improved a bit.

Cyan4973 commented on lib/compress/zstd_compress.c:
```diff
@@ -2626,6 +2657,166 @@ ZSTD_compressSequences_internal(seqStore_t* seqStorePtr,
+    }
+
+    /* Scan input and build symbol stats */
+    {   size_t largest = HIST_count(count, &maxSymbolValue, (const BYTE*)src, srcSize);
```
nit : `const largest`

ephiepark committed:
Libfuzzer to test this feature

Cyan4973 commented on lib/compress/zstd_compress_internal.h:
```diff
@@ -81,6 +81,25 @@ typedef struct {
     ZSTD_fseCTables_t fse;
 } ZSTD_entropyCTables_t;
 
+typedef struct {
```
Try to define and document "metadata" in this context, as its a fairly generic term.

ephiepark replied:
Sounds good!

Cyan4973 commented:
One completely opened question, 
while evaluating the amount of changes within `ZSTD_compress.c`, I was wondering : would it be preferable to organize the block splitter code into its own source file ?

ephiepark committed:
Update minimum possible size for sequences bitstream

ephiepark commented:
Moving to its own source file makes sense. Let me do that.

ephiepark committed:
Restructure the source files

Cyan4973 commented:
I guess the new file is `compress_advanced.{c,h} ?

I believe this name is not specific enough: it could contain anything labelled "advanced", of which `zstd` is already plentiful.

Try something which suggests more precisely what you are doing, such as `ZSTD_compress_smallBlocks` for examples, or anything of your liking that give a reviewer a chance to guess what this file contains.

Cyan4973 commented on programs/zstdcli.c:
```diff
@@ -141,6 +141,7 @@ static int usage_advanced(const char* programName)
     DISPLAY( "--long[=#]: enable long distance matching with given window log (default: %u)\n", g_defaultMaxWindowLog);
     DISPLAY( "--fast[=#]: switch to ultra fast compression level (default: %u)\n", 1);
     DISPLAY( "--adapt : dynamically adapt compression level to I/O conditions \n");
+    DISPLAY( "--target-compressed-block-size=# : try to make compressed block size to be target-compressed-block-size \n");
```
nit: try to shorten explanation, so that it fits into 80-columns.

Cyan4973 commented on lib/compress/zstd_compress_sequences.h:
```diff
@@ -0,0 +1,420 @@
+ * If x == 0: Return 0
+ * Else: Return floor(-log2(x / 256) * 256)
+ */
+static unsigned const kInverseProbabilityLog256[256] = {
```
minor note on code size :
this will create an instance of the table in each file where it's included and used.

ephiepark replied:
I see. I can create .c file for zstd_compress_sequences and move this variable there. I have a few questions though.

1. Do you think my approach of organizing code for zstd_compress_sequences and zstd_compress_literals make sense?

2. What are the rules of thumb for writing the function implementation within header file vs. just writing function signature in header file and implementing it in c file? In zstd_compress_interal.h I see some functions are implemented directed in the header file (like `ZSTD_fWeight()`) while other functions like `ZSTD_resetSeqStore()` is only declared in header file and implemented in c file.

Cyan4973 commented on lib/compress/zstd_compress_advanced.h:
```diff
@@ -0,0 +1,56 @@
+    symbolEncodingType_e hType;
+    BYTE hufDesBuffer[500];
+    size_t hufDesSize;
+} ZSTD_hufCTablesMetadata_t;
```
It only matters to expose these definitions if they are going to be used by the "user" (the source file which `#include` this header).

Typically, it would be necessary if this type was required to invoke one of the following functions.

But that's not the case. Apparently these types are only useful within `zstd_compress_advanced.c`, so, following the principle of leanest scope, that's where they should be declared.

ephiepark commented:
I see. I named it to be zstd_compress_advanced to see if it make sense to put all "advanced" features into that file but if there are already plentiful of them, I think it makes sense not to have a kitchen sink file and have one for each. How about zstd_compress_superblock.{c, h} to make it match the entry function name?

Cyan4973 commented:
> How about zstd_compress_superblock.{c, h} to make it match the entry function name?

Sounds great !

ephiepark commented:
@Cyan4973 I have updated the pull request so that the implementation of the functions are specified in c files and header files only have function signatures that are used by users of the header files. 

Side effects of these changes are:
- functions like `ZSTD_noCompressLiterals`, `ZSTD_compressRleLiteralsBlock`, `ZSTD_selectEncodingType`, `ZSTD_buildCTable`, `ZSTD_encodeSequences` are no longer static functions.
- `ZSTD_selectEncodingType`, `ZSTD_buildCTable` used to be marked with `MEM_STATIC` which I believe is meant to inline the functions. However, in the process of moving implementation to a c file, I removed `MEM_STATIC` so they won't be inlined anymore. 

Are these side effects okay?

Cyan4973 commented:
`MEM_STATIC` is [defined in `mem.h`](https://github.com/facebook/zstd/blob/dev/lib/common/mem.h#L26). Using only "standard C", it is equivalent to `static inline`.

One issue with `inline` is that this keyword is not defined for `C90`. So it must be used cautiously (with macro guards). That's what the macro is doing.

But I think the real objective was the `__attribute__((unused))`.

I'm not sure why it was used on these 2 functions. `MEM_STATIC` is rather used for functions which are defined directly in headers (see paragraph below), for complex reasons related to compiler and warnings. But these 2 functions were not part of a header file, so I have no reasonable idea.

When something really _has to_ be inlined, we generally use `FORCE_INLINE_TEMPLATE`, which is like an order that the compiler can't refuse. Since it was not used, it implies that inlining wasn't that important.

For your specific case :
- `ZSTD_buildCTable` : I don't think there is any reason to inline that
- `ZSTD_selectEncodingType` :  I'm not sure. I have not written this function. When I look at its body, and its invocations, I suspect inlining is not important : the function is large, and I don't notice any key simplification that would derive from using a constant parameter. So it's probably fine with `static` too.

From a more general perspective, it's recommended to benchmark and compare before / after when a change can impact performance.

----------------

The problem of `static` functions is that, when defined but ___not used___, some compilers will issue a warning ("unused function"). This is generally a good thing, as it allows to catch forgotten functions that are no longer useful. It will happen with our standard set of compilers and flags. However, when included from a `*.h` file, the `static` function might be sometimes invoked, sometimes not. And since the "unused function" warning is local, it will be triggered, even though the function is actually useful from other contexts.

`MEM_STATIC` is our complex macro to pass that. It means "do not warn if this static function is unused". `inline` is another way to request the same thing (the function is defined because it needs to be inlined, _hence_ do not warn if it's not invoked), but `inline` is merely a hint, its impact on compiler warnings is not guaranteed, hence `__attribute__((unused))` is safer when available (`gcc` and `clang` mostly).

ephiepark commented:
@Cyan4973 Thanks for the thorough explanation! So, it seems like there is no obvious issue from this changes. I will start cleaning up zstd_compress_superblock code. After that, I will run benchmarks to see if there was any noticeable performance regression introduced by my changes.

ephiepark committed:
Cleaning up zstd_compress_superblock

Cyan4973 commented on lib/compress/zstd_compress.c:
```diff
@@ -2820,6 +2393,39 @@ static size_t ZSTD_compressBlock_internal(ZSTD_CCtx* zc,
+    DEBUGLOG(5, "ZSTD_compressSuperBlock (dstCapacity=%u, dictLimit=%u, nextToUpdate=%u, srcSize=%zu)",
+                (unsigned)dstCapacity, (unsigned)zc->blockState.matchState.window.dictLimit, (unsigned)zc->blockState.matchState.nextToUpdate, srcSize);
+
+    {   size_t bss = ZSTD_buildSeqStore(zc, src, srcSize);
```
nit: `const bss`

Cyan4973 commented on lib/compress/zstd_compress_literals.h:
```diff
@@ -0,0 +1,18 @@
+ */
+
+#ifndef ZSTD_COMPRESS_LITERALS_H
+#define ZSTD_COMPRESS_LITERALS_H
```
`zstd_compress_literals.c` only contains "special cases", RLE and noCompress,
but the "normal" huffman compression is elsewhere.
Why ?

ephiepark replied:
It was because ZSTD_compressLiterals depends on ZSTD_minGain which isn't literals specific. In order to prevent circular dependency, I moved ZSTD_minGain to zstd_compress_internal.h at which point I ran into static function not being used warning. 

I think can do the similar approach with MEM_STATIC to suppress the warning. Let me do that!

Cyan4973 commented on lib/compress/zstd_compress_literals.c:
```diff
@@ -0,0 +1,67 @@
+ /*-*************************************
+ *  Dependencies
+ ***************************************/
+#include "zstd_internal.h"
```
Could you detail in code comment what you need from these `#include` ?

Cyan4973 commented on lib/compress/zstd_compress_sequences.h:
```diff
@@ -0,0 +1,48 @@
+#ifndef ZSTD_COMPRESS_SEQUENCES_H
+#define ZSTD_COMPRESS_SEQUENCES_H
+
+// #define FSE_STATIC_LINKING_ONLY   /* FSE_encodeSymbol */ TODO is this necessary?
```
`FSE_STATIC_LINKING_ONLY` might be necessary inside the `*.c` implementation, rather than the `*.h` declaration.

In general, we try to reduce the nb of dependencies to a minimum.
This is especially important in header files, so we try to justify each dependency with a code comment.

Cyan4973 commented on lib/compress/zstd_compress_superblock.h:
```diff
@@ -0,0 +1,36 @@
+*  Dependencies
+***************************************/
+
+#include "zstd_internal.h"
```
why do you need this `#include` ?

Cyan4973 commented on lib/compress/zstd_compress_superblock.h:
```diff
@@ -0,0 +1,36 @@
+#include "zstd_internal.h"
+
+
+// #if defined (__cplusplus) TODO are these necessary?
```
I think it's not necessary, because it's not a user-accessible interface.

Cyan4973 commented on lib/compress/zstd_compress_superblock.c:
```diff
@@ -0,0 +1,567 @@
+ *  This metadata is populated in ZSTD_buildSuperBlockEntropy_literal() */
+typedef struct {
+    symbolEncodingType_e hType;
+    BYTE hufDesBuffer[500]; // TODO give name to this value
```
Probably too generous a budget. Not a blocker. Something to optimize a bit later.

Cyan4973 commented on lib/compress/zstd_compress_superblock.c:
```diff
@@ -0,0 +1,567 @@
+    symbolEncodingType_e llType;
+    symbolEncodingType_e ofType;
+    symbolEncodingType_e mlType;
+    BYTE fseTablesBuffer[500]; // TODO give name to this value
```
Probably too generous a budget. Not a blocker. Something to optimize a bit later.

Cyan4973 commented on lib/compress/zstd_compress_superblock.c:
```diff
@@ -0,0 +1,567 @@
+        FORWARD_IF_ERROR(maxBits);
+        huffLog = (U32)maxBits;
+        hufMetadata->hType = set_compressed;
+        return HUF_writeCTable(hufMetadata->hufDesBuffer, sizeof(hufMetadata->hufDesBuffer),
```
At this point, you both write the literals' section header, and decide to go for huffman compression of literals.
All "obvious" corner cases have already been filtered out, but it's still possible that the sum of `header size + compressed literals` is ` > uncompressed literals`, especially when nb of literals is small.

In the "normal" implementation, this is checked after compression, on compressed data itself.
But here, compression isn't happening, so the check has simply been suppressed.

Suggestion : ensure that compression of literals will at least produce _some_ gains.
It's possible to quickly evaluate the cost of compressed literals just from the histogram and the `CTable` : just sum the bits. I suspect we already have a function for this evaluation. If not, it should be easy to create.

Stretch goal : consider the cost of headers. I believe uncompressed literals have a small advantage regarding header size. Compressed blocks must over-match that. Don't try to be precise, a loose or slightly generous estimation will do fine enough.

ephiepark replied:
For now, I will focus on getting the existing logic reviewed, and tackle this with different commit.

Cyan4973 commented on lib/compress/zstd_compress_superblock.c:
```diff
@@ -0,0 +1,567 @@
+    BYTE* const oend = ostart + sizeof(fseMetadata->fseTablesBuffer);
+    BYTE* op = ostart;
+    U32 LLtype, Offtype, MLtype;   /* compressed, raw or rle */
+    BYTE workspace[HIST_WKSP_SIZE];
```
While this is convenient, this kind of stack usage can generate problems in certain setup where stack space is scarce (kernel development is a prime example).
To avoid that, we try to borrow the workspace from the allocated context.
Consider this strategy.
If too complex, comment clearly this section, so that it can be fixed later, when needed (@terrelln plans to release an updated version of `zstd` for Linux Kernel for example).

Cyan4973 commented on lib/compress/zstd_compress_superblock.c:
```diff
@@ -0,0 +1,567 @@
+    }
+
+    // TODO bmi2
+    {   size_t cSize = singleStream ? HUF_compress1X_usingCTable(op, oend-op, literals, litSize, hufTable)
```
nit: `const cSize`

Cyan4973 commented on lib/compress/zstd_compress_superblock.c:
```diff
@@ -0,0 +1,567 @@
+    BYTE* const oend = ostart + dstCapacity;
+    BYTE* op = ostart;
+    BYTE* seqHead;
+    U32 LLtype, Offtype, MLtype;
```
I think the scope of these variables can be sharply reduced. See comments below.

Cyan4973 commented on lib/compress/zstd_compress_superblock.c:
```diff
@@ -0,0 +1,567 @@
+
+    /* seqHead : flags for FSE encoding type */
+    seqHead = op++;
+    *seqHead = (BYTE)((LLtype<<6) + (Offtype<<4) + (MLtype<<2));
```
If I understand correctly, all `*type` values are `set_repeat` at this point.

Cyan4973 commented on lib/compress/zstd_compress_superblock.c:
```diff
@@ -0,0 +1,567 @@
+
+    if (writeEntropy) {
+        DEBUGLOG(5, "ZSTD_compressSubBlock_sequences (fseTablesSize=%zu)", fseMetadata->fseTablesSize);
+        LLtype = fseMetadata->llType;
```
`*type` is only used in the construction of `*seqHead`.
There is no need for this variable lifetime to go beyond this block.

Cyan4973 commented on lib/compress/zstd_compress_superblock.c:
```diff
@@ -0,0 +1,567 @@
+        DEBUGLOG(5, "ZSTD_compressSubBlock_sequences (bitstreamSize=%zu)", bitstreamSize);
+    }
+
+    /* zstd versions <= x.x.x mistakenly report error when
```
x.x.x is `1.4.0`

Cyan4973 commented on lib/compress/zstd_compress_superblock.c:
```diff
@@ -0,0 +1,567 @@
+        memcpy(op, fseMetadata->fseTablesBuffer, fseMetadata->fseTablesSize);
+        op += fseMetadata->fseTablesSize;
+    }
+
```
for clarity, I would rather write `*seqHead` with `set_repeat` here, in an `else` statement.

Cyan4973 commented on lib/compress/zstd_compress_superblock.c:
```diff
@@ -0,0 +1,567 @@
+        seqCount++;
+        cBlockSizeEstimate = sizeof(seqDef) * seqCount + litSize;
+        if (cBlockSizeEstimate > targetCBlockSize || lastSequence) {
+            size_t decompressedSize = ZSTD_seqDecompressedSize(sp, seqCount, litSize);
```
nit : use `const`

Cyan4973 commented:
I like the code structure. It's very clear, it makes it easy to review.

The whole patch is large though, which makes it difficult for review to feel certain that all issues are detected.
Automated tests are therefore important to strengthen the confidence that the patch doesn't introduce new problems in existing scenarios.
Adding your path to `tests/fuzz` is a good step in this direction.

terrelln commented on lib/compress/zstd_compress_superblock.c:
```diff
@@ -0,0 +1,567 @@
+                                                  ZSTD_hufCTables_t* nextHuf,
+                                                  ZSTD_hufCTablesMetadata_t* hufMetadata)
+{
+    unsigned count[HUF_SYMBOLVALUE_MAX + 1];
```
We should pass a workspace in. This is 1 KB of stack space. On some ARM devices in the kernel we have 8 KB of stack total.

terrelln commented on lib/compress/zstd_compress_superblock.c:
```diff
@@ -0,0 +1,567 @@
+                                                    ZSTD_fseCTablesMetadata_t* fseMetadata)
+{
+    ZSTD_strategy const strategy = cctxParams->cParams.strategy;
+    unsigned count[MaxSeq+1];
```
This should also go in the workspace.

ephiepark replied:
I have replace everywhere with workspace by looking at zstd_compress to see which workspace is used for what purposes. This one, I am not sure which workspace to use since zstd_compress.c also uses stack memory for this purpose. This one is small though. Assuming unsigned is 4 byte this array is going to be 4 X 53 = 212 bytes.

Cyan4973 replied:
212 bytes seems reasonable to me.

If some very strict stack usage requirements mandate to use less, I believe it's better to make it a separate topic.

ephiepark replied:
I think count is allocated in stack because buildCTable() uses workspace and there is a danger of overwriting count. But I think buildCTable's workspace need is small enough that I think workspace can be used for both count[] and buildCTable. Let me try that. I am not sure how to make it future proof though. (Should I even worry about workspace size need changing over time?) I will need some feedback.

terrelln commented on lib/compress/zstd_compress_superblock.h:
```diff
@@ -0,0 +1,36 @@
+/* ZSTD_compressSuperBlock_internal() :
+ * Used to compress a super block when targetCBlockSize is being used.
+ * The given block will be compressed into multiple sub blocks that are around targetCBlockSize. */
+size_t ZSTD_compressSuperBlock_internal(ZSTD_CCtx* zc,
```
Can we call this `ZSTD_compressSuperBlock()`?

ephiepark replied:
There is a ZSTD_compressSuperBlock() in zstd_compress.c which calls this function. Let me think about their names.
